### PR TITLE
Bug 1687644 - Implement the external API for consumers

### PR DESCRIPTION
**Pull request by brizental**
Now that we can acutally measure it, for reference:

- The webext build is 46K.
- The qt build is 44K.

### CODE REVIEW

**Comment by brizental** [comment]
The Qt error seems unrelated. I will verify, but don't feel that it blocks the review :)

File: package.json [hunk 1]
```diff
@@ -9,9 +9,9 @@
     "lint": "eslint . --ext .ts,.js,.json --max-warnings=0",
     "fix": "eslint . --ext .ts,.js,.json --fix",
     "build:webext": "webpack --config webpack.config.webext.js --mode production",
-    "dev:webext": "webpack --watch --config webpack.config.webext.js --mode development",
+    "dev:webext": "webpack --watch --config webpack.config.webext.js --mode development --devtool inline-source-map",
```
Comment (Dexterp37) [inline package.json:12]: Nice move :) I'll borrow this for Rally.

File: src/index.ts [hunk 2]
```diff
@@ -2,25 +2,81 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+import Glean from "glean";
+import Configuration from "config";
+
+// Private stuff to export.
```
Comment (Dexterp37) [inline src/index.ts:8]:
```suggestion
// Private Glean types to export.
```

File: src/storage/utils.ts [hunk 3]
```diff
@@ -23,7 +23,10 @@ export function getValueFromNestedObject(obj: JSONObject, index: StorageIndex):
   let target: JSONValue = obj;
   for (const key of index) {
     if (isObject(target) && key in target) {
-      target = target[key];
+      const temp: unknown = target[key];
+      if (isJSONValue(temp)) {
```
Comment (Dexterp37) [inline src/storage/utils.ts:27]: Did you stumble upon non-JSON values being stored?

File: src/upload/uploader/index.ts [hunk 4]
```diff
@@ -34,7 +34,11 @@ export abstract class Uploader {
 //
 // TODO: remove this comment once Bug 1687516 is resolved.
 class MockUploader extends Uploader {
-  post(): Promise<UploadResult> {
+  // If we don't include these arguments here, the typescript checker will raise
+  // `TS2554: Expected 0 arguments, but got 3.`.
+  //
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Comment (Dexterp37) [inline src/upload/uploader/index.ts:40]: It makes sense since the abstract class has them. Do we need to specify this?

File: src/storage/weak.ts [hunk 5]
```diff
@@ -14,7 +14,11 @@ import { JSONObject, JSONValue } from "utils";
 class WeakStore implements Store {
   private store: JSONObject;
 
-  constructor() {
+  // If we don't include these arguments here, the typescript checker will raise
+  // `TS2554: Expected 0 arguments, but got 1.`.
+  //
```
Comment (Dexterp37) [inline src/storage/weak.ts:19]: It makes sense since the abstract class has it. Do we need to specify this?

**Review by Dexterp37 — CHANGES_REQUESTED** [review]

File: src/upload/uploader/index.ts (hunk 4 above)
Comment (brizental) [reply src/upload/uploader/index.ts:40]: Eh, you are right. I'll remove the comment.

File: src/index.ts (hunk 2 above)
Comment (brizental) [reply src/index.ts:8]: Heh, much nicer.

**Comment by brizental** [comment]
The Qt problems were actually related. We were using some browser APIs namely URL and crypto, which I didn't realize before. That was fixed. This is ready for a new look.

**Review by Dexterp37 — APPROVED** [review]

**Pull request merged**